quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/services/resources/QuarkusWelcomeResource.java

```diff
@@ -16,209 +16,24 @@
-        try {
-            Theme theme = getTheme();
-
-            if(Objects.isNull(theme)) {
-                Log.error("Theme is null please check the \"--spi-theme-default\" parameter");
-                errorMessage = "The theme is null";
-                ResponseBuilder rb = Response.status(Status.BAD_REQUEST)
-                        .entity(errorMessage)
-                        .cacheControl(CacheControlUtil.noCache());
-                return rb.build();
-            }
-
-            Map<String, Object> map = new HashMap<>();
-
-            map.put("adminConsoleEnabled", isAdminConsoleEnabled());
-            map.put("productName", Version.NAME);
-
-            map.put("properties", theme.getProperties());
-            map.put("adminUrl", session.getContext().getUri(UrlType.ADMIN).getBaseUriBuilder().path("/admin/").build());
-
-            map.put("resourcesPath", "resources/" + Version.RESOURCES_VERSION + "/" + theme.getType().toString().toLowerCase() +"/" + theme.getName());
-            map.put("resourcesCommonPath", "resources/" + Version.RESOURCES_VERSION + "/common/keycloak");
+public class QuarkusWelcomeResource extends WelcomeResource {
 
-            boolean bootstrap = shouldBootstrap();
-            map.put("bootstrap", bootstrap);
-            if (bootstrap) {
-                boolean isLocal = isLocal();
-                map.put("localUser", isLocal);
-
-                String localAdminUrl = getLocalAdminUrl();
-
-                map.put("localAdminUrl", localAdminUrl);
-                map.put("adminUserCreationMessage", "or set the environment variables KEYCLOAK_ADMIN and KEYCLOAK_ADMIN_PASSWORD before starting the server");
-
-                if (isLocal) {
-                    String stateChecker = setCsrfCookie();
-                    map.put("stateChecker", stateChecker);
-                }
-            }
-            if (successMessage != null) {
-                map.put("successMessage", successMessage);
-            }
-            if (errorMessage != null) {
-                map.put("errorMessage", errorMessage);
-            }
-            FreeMarkerProvider freeMarkerUtil = session.getProvider(FreeMarkerProvider.class);
-            String result = freeMarkerUtil.processTemplate(map, "index.ftl", theme);
-
-            ResponseBuilder rb = Response.status(errorMessage == null ? Status.OK : Status.BAD_REQUEST)
-                    .entity(result)
-                    .cacheControl(CacheControlUtil.noCache());
-            return rb.build();
-        } catch (Exception e) {
-            throw new WebApplicationException(e, Response.Status.INTERNAL_SERVER_ERROR);
-        }
+    @Override
+    protected String getAdminCreationMessage() {
+        return "or set the environment variables KEYCLOAK_ADMIN and KEYCLOAK_ADMIN_PASSWORD before starting the server";
     }
 
-    private String getLocalAdminUrl() {
+    @Override
+    protected String getLocalAdminUrl() {
```

Comment:
I think we can remove this method too and then finally remove the class. For that, we can add a URL type `org.keycloak.urls.UrlType.LOCAL_ADMIN` and resolve this URL from the hostname provider itself.

If you want I can give it a try.

Reply:
> If you want I can give it a try.

Sounds good.  If the local admin url is used in other places, it would be great to have it consolidated.

Reply:
That is the thing ... This URL type would be used only here.

However, having it resolved from the hostname provider should make life a lot easier because we don't need to deal with the differences around the runtime (e.g.: undertow vs quarkus). And ultimately removing this class completely.

In the [hostname provider](https://github.com/keycloak/keycloak/blob/main/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/hostname/DefaultHostnameProvider.java) we already have similar code to resolve the scheme and the port.

Reply:
> In the [hostname provider](https://github.com/keycloak/keycloak/blob/main/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/hostname/DefaultHostnameProvider.java) we already have similar code to resolve the scheme and the port.

I've updated the pr with a simplistic refactoring that moves the local admin to the hostnameprovider.  However I didn't want to guess too much about what could be consolidated with the quarkus DefaultHostnameProvider, since the local admin construction logic was subtly different.

Is this along the lines of what you are thinking?

Reply:
Thanks, @shawkins.